### PR TITLE
p2p/discover: sha3-based node distance

### PR DESCRIPTION
**Pull request by fjl**
This PR builds on the endpoint format changes (#743) mostly to reduce merge trouble later.

### CODE REVIEW

File: p2p/discover/node.go [hunk 1]
```diff
@@ -81,36 +84,50 @@ func (n *Node) String() string {
 //
 //    enode://<hex node id>@10.3.58.6:30303?discport=30301
 func ParseNode(rawurl string) (*Node, error) {
-	var n Node
+	var (
+		id               NodeID
+		ip               net.IP
+		tcpPort, udpPort uint64
```
Comment (Gustav-Simonsson) [inline p2p/discover/node.go]: perhaps use uint16 for consistency with Node.UDP/TCP

Comment (Gustav-Simonsson) [inline p2p/discover/node.go]: Ah, nevermind, I see it's used due to strconv.ParseUInt returning uint64. nm

**Comment by fjl** [comment]
This needs a rebase on #743.

**Comment by fjl** [comment]
Has been rebased.

**Pull request merged**